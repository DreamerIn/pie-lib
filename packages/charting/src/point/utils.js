import debug from 'debug';
import omitBy from 'lodash/omitBy';
import isUndefined from 'lodash/isUndefined';

const log = debug('pie-lib:charting:point:utils');

/**
 * Get the index of the point in the points array
 * @param {Point[]} points
 * @param {Point} p
 */
export const pointIndex = (points, p) => {
  return (points || []).findIndex(s => {
    return s.x === p.x && s.y === p.y;
  });
};

export const hasPoint = (points, p) => pointIndex(points, p) !== -1;

export const trim = point => ({
  x: point.x,
  y: point.y,
  label: point.label
});

export const removePoints = (points, toRemove) => {
  log('[removePoints] points: ', points, toRemove);
  toRemove = Array.isArray(toRemove) ? toRemove : [toRemove];

  const out = points.filter(p => !hasPoint(toRemove, p));

  log('[removePoints] points: ', out);
  return out;
};

export const firstAvailableLabel = (points, labels) => {
  const existing = points.map(p => p.label);
  log('points: ', points);
  const filtered = labels.filter(l => !existing.some(e => e === l));
  log('existing: ', existing, 'filtered: ', filtered);
  return filtered.length > 0 ? filtered[0] : '';
};

export const addPoint = (points, p, labels) => {
  labels = labels || [];
  if (!hasPoint(points, p)) {
    const out = points.slice();
    const label = firstAvailableLabel(points, labels);
    out.push(Object.assign({}, p, { label }));
    return out;
  } else {
    return points;
  }
};

export const bounds = (p, domain, range) => {
  return {
    left: domain.min - p.x,
    right: Math.abs(p.x - domain.max),
    top: Math.abs(p.y - range.max),
    bottom: range.min - p.y
  };
};

export const swap = (arr, from, to, getIndex) => {
  log('[swap] arr: ', arr, 'from: ', from, 'to: ', to);

  const toIndex = getIndex(arr, to);

  log('[swap] toIndex: ', toIndex);

  if (toIndex !== -1) {
    //prevent moving one point on top of another.
    return arr;
  }

  const fromIndex = getIndex(arr, from);

  log('[swap] fromIndex: ', fromIndex);

  if (fromIndex === -1) {
    //cant move a point that isn't in the points array
    return arr;
  }

  const f = arr[fromIndex];
  const replacement = Object.assign({}, f, omitBy(to, isUndefined));
  const out = arr.slice();
  out.splice(fromIndex, 1, replacement); //{ ...to, label: f.label });
  return out;
};

/**
 * swap a point 'from', with point 'to'.
 * Finds the point in the array and updates it to equal the 'to' point.
 * @param {Point[]} points
 * @param {Point} from
 * @param {Point} to
 */

export const swapPoint = (points, from, to) => {
  return swap(points, from, to, pointIndex);
};
