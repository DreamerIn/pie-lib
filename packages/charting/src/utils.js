import { scaleBand, scalePoint } from '@vx/scale';
import { utils } from '@pie-lib/plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const bandKey = (d, index) => `${index}-${d.label || '-'}`;

export const dataToXBand = (scaleX, data, width, type) => {
  switch (type) {
    case 'bar':
    case 'dotPlot':
    case 'linePlot':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0.2
      });
    case 'histogram':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0
      });
    case 'line':
      return scalePoint({
        domain: data && data.map(bandKey),
        rangeRound: [0, width]
      });
    default:
      return scaleBand({
        range: [0, width],
        domain: data && data.map(bandKey),
        padding: 0
      });
  }
};

export const getTickValues = (prop = {}) => {
  const tickValues = [];
  let tickVal = prop.min;

  while (tickVal <= prop.max) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal + prop.step) * 100) / 100;
  }

  return tickValues;
};

export const getDomainAndRangeByChartType = (domain, range, chartType) => {
  const { min, max } = range || {};
  let { step } = range || {};
  const numberMax = parseFloat(max);

  if (!step) {
    if (numberMax < 20) {
      step = 1;
    } else if (numberMax >= 20 && numberMax < 100) {
      step = 5;
    } else {
      step = 10;
    }
  }

  switch (chartType) {
    // if chart is dot plot or line plot, we should ignore step and make sure that min & max are integer values
    case 'dotPlot':
    case 'linePlot':
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1
        },
        range: {
          ...range,
          min: parseInt(min, 10),
          max: parseInt(max, 10),
          step: 1
        }
      };
    default:
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1
        },
        range: {
          ...range,
          step
        }
      };
  }
};

export const getGridLinesAndAxisByChartType = (range, chartType) => {
  switch (chartType) {
    case 'line':
      return {
        verticalLines: undefined,
        horizontalLines: getTickValues(range),
        leftAxis: true
      };
    case 'dotPlot':
    case 'linePlot':
      return {
        verticalLines: [],
        horizontalLines: [],
        leftAxis: false
      };
    default:
      return {
        verticalLines: [],
        horizontalLines: getTickValues(range),
        leftAxis: true
      };
  }
};

export const getRotateAngle = barWidth => {
  if (barWidth < 30) {
    return 75;
  }

  if (barWidth < 40) {
    return 45;
  }

  if (barWidth < 60) {
    return 25;
  }

  return 0;
};

export const getTopPadding = barWidth => {
  if (barWidth < 30) {
    return 50;
  }

  if (barWidth < 40) {
    return 30;
  }

  if (barWidth < 60) {
    return 15;
  }

  return 0;
};
