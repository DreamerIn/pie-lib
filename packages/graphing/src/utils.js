import head from 'lodash/head';
import tail from 'lodash/tail';
import { utils } from '@pie-lib/plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

// export { * from utils}
export const polygonToArea = points => {
  const h = head(points);
  const area = {
    left: h.x,
    top: h.y,
    bottom: h.y,
    right: h.x
  };
  return tail(points).reduce((a, p) => {
    a.left = Math.min(a.left, p.x);
    a.top = Math.max(a.top, p.y);
    a.bottom = Math.min(a.bottom, p.y);
    a.right = Math.max(a.right, p.x);
    return a;
  }, area);
};

export const lineToArea = (from, to) => {
  const left = Math.min(from.x, to.x);
  const top = Math.max(from.y, to.y);
  const bottom = Math.min(from.y, to.y);
  const right = Math.max(from.x, to.x);
  return { left, top, bottom, right };
};
