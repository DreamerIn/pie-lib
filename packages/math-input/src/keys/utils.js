export const mkSet = category => o => ({ ...o, category });

export const toArray = o => Object.keys(o).map(k => o[k]);

export const toKey = category => v => {
  if (typeof v === 'string') {
    return {
      name: v,
      label: v,
      category,
      write: v
    };
  } else {
    return {
      ...v,
      category
    };
  }
};
