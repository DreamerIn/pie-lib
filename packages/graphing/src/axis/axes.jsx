import React from 'react';
import { Axis } from '@vx/axis';
import { types, utils } from '@pie-lib/plot';
import PropTypes from 'prop-types';
import Arrow from './arrow';
import { withStyles } from '@material-ui/core';
import isEqual from 'lodash/isEqual';
import { getTickValues } from '../utils';

const AxisPropTypes = {
  includeArrows: PropTypes.bool
};
const AxisDefaultProps = {
  includeArrows: true
};

const axisStyles = theme => ({
  line: {
    stroke: theme.palette.primary.main,
    strokeWidth: 5
  },
  arrow: {
    fill: theme.palette.primary.main
  },
  tick: {
    fill: theme.palette.primary.main,
    '& > line': {
      stroke: theme.palette.primary.main
    }
  },
  axisLabelHolder: {
    padding: 0,
    margin: 0,
    textAlign: 'center',
    '* > *': {
      margin: 0,
      padding: 0
    }
  }
});

const tickLabelStyles = {
  fontFamily: 'Roboto',
  fontSize: '14px',
  cursor: 'inherit'
};

export class RawXAxis extends React.Component {
  static propTypes = {
    ...AxisPropTypes,
    classes: PropTypes.object,
    graphProps: types.GraphPropsType.isRequired
  };
  static defaultProps = AxisDefaultProps;

  shouldComponentUpdate(nextProps) {
    const { data } = this.props;
    return (
      !utils.isDomainRangeEqual(this.props.graphProps, nextProps.graphProps) ||
      !isEqual(data, nextProps.data)
    );
  }
  render() {
    const { includeArrows, classes, graphProps } = this.props;
    const { scale, domain, size } = graphProps;
    const columnTicksValues = getTickValues({ ...domain, step: domain.labelStep });

    return (
      <React.Fragment>
        <Axis
          axisLineClassName={classes.line}
          scale={scale.x}
          top={scale.y(0)}
          left={0}
          label={domain.label}
          tickClassName={classes.tick}
          tickFormat={value => value}
          tickLabelProps={label => ({
            ...tickLabelStyles,
            y: 25,
            dx: label === '0' ? 6 : -4
          })}
          // Having 0 as a number in columnTicksValues does not make 0 to show up
          // so we use this trick:
          tickValues={['0', ...columnTicksValues]}
        />
        {includeArrows && (
          <Arrow direction="left" x={domain.min} y={0} className={classes.arrow} scale={scale} />
        )}
        {includeArrows && (
          <Arrow direction="right" x={domain.max} y={0} className={classes.arrow} scale={scale} />
        )}
        {domain.axisLabel && (
          <foreignObject x={size.width + 10} y={scale.y(0) - 10} width={100} height={20}>
            <div dangerouslySetInnerHTML={{ __html: domain.axisLabel }} />
          </foreignObject>
        )}
      </React.Fragment>
    );
  }
}

const XAxis = withStyles(axisStyles)(RawXAxis);

export class RawYAxis extends React.Component {
  static propTypes = {
    ...AxisPropTypes,
    graphProps: types.GraphPropsType.isRequired
  };
  static defaultProps = AxisDefaultProps;
  shouldComponentUpdate(nextProps) {
    const { data } = this.props;
    return (
      !utils.isDomainRangeEqual(this.props.graphProps, nextProps.graphProps) ||
      !isEqual(data, nextProps.data)
    );
  }
  render() {
    const { classes, includeArrows, graphProps } = this.props;
    const { scale, range, size } = graphProps;
    const rowTickValues = getTickValues({ ...range, step: range.labelStep });

    return (
      <React.Fragment>
        <Axis
          axisLineClassName={classes.line}
          orientation={'left'}
          scale={scale.y}
          top={0}
          height={size.height}
          left={scale.x(0)}
          label={range.label}
          tickLength={10}
          tickClassName={classes.tick}
          tickFormat={value => value}
          tickLabelProps={value => {
            const digits = value.toLocaleString().length || 1;

            return {
              ...tickLabelStyles,
              dy: 4,
              dx: -8 - digits * 5
            };
          }}
          hideZero={true}
          tickTextAnchor={'bottom'}
          tickValues={rowTickValues}
        />
        {includeArrows && (
          <Arrow direction="down" x={0} y={range.min} className={classes.arrow} scale={scale} />
        )}
        {includeArrows && (
          <Arrow direction="up" x={0} y={range.max} className={classes.arrow} scale={scale} />
        )}
        {range.axisLabel && (
          <foreignObject x={scale.x(0) - 50} y={-25} width="100" height="20">
            <div
              dangerouslySetInnerHTML={{ __html: range.axisLabel }}
              className={classes.axisLabelHolder}
            />
          </foreignObject>
        )}
      </React.Fragment>
    );
  }
}
const YAxis = withStyles(axisStyles)(RawYAxis);

export default class Axes extends React.Component {
  static propTypes = AxisPropTypes;
  static defaultProps = AxisDefaultProps;
  render() {
    return (
      <React.Fragment>
        <XAxis {...this.props} />
        <YAxis {...this.props} />
      </React.Fragment>
    );
  }
}
