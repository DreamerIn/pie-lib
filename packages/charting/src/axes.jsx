import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { utils, types } from '@pie-lib/plot';
import { AxisLeft, AxisBottom } from '@vx/axis';
import { dataToXBand, getTickValues } from './utils';

class RawChartAxes extends React.Component {
  static propTypes = {
    bottomScale: PropTypes.func,
    classes: PropTypes.object.isRequired,
    data: PropTypes.array,
    graphProps: types.GraphPropsType.isRequired
  };

  render() {
    const { classes, data, graphProps, type } = this.props;
    const { scale, range, domain, size } = graphProps;
    const xBand = dataToXBand(scale.x, data, size.width, type);
    const bottomScale = xBand.rangeRound([0, size.width]);
    const rowTickValues = getTickValues({ ...range, step: range.labelStep });

    return (
      <React.Fragment>
        <AxisLeft
          scale={scale.y}
          className={classes.axis}
          axisLineClassName={classes.axisLine}
          tickLength={10}
          tickClassName={classes.tick}
          tickFormat={value => value}
          label={range.label}
          labelClassName={classes.axisLabel}
          tickValues={rowTickValues}
          tickLabelProps={value => {
            const digits = value.toLocaleString().length || 1;

            return {
              dy: 4,
              dx: -10 - digits * 5
            };
          }}
        />
        <AxisBottom
          axisLineClassName={classes.axisLine}
          labelClassName={classes.axisLabel}
          tickClassName={classes.tick}
          scale={bottomScale}
          tickComponent={props => {
            return (
              <text x={props.x} y={props.y} className={classes.tick}>
                {props.formattedValue.split('-')[1]}
              </text>
            );
          }}
          label={domain.label}
          top={scale.y(range.min)}
          textLabelProps={() => ({ textAnchor: 'middle' })}
          tickFormat={count => count}
        />
      </React.Fragment>
    );
  }
}

const ChartAxes = withStyles(theme => ({
  axisLabel: {
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    fill: theme.palette.secondary.main
  },
  axis: {
    stroke: theme.palette.primary.dark,
    strokeWidth: 2
  },
  axisLine: {
    stroke: theme.palette.primary.dark,
    strokeWidth: 2
  },
  tick: {
    '& > line': {
      stroke: theme.palette.primary.dark,
      strokeWidth: 2
    },
    fill: theme.palette.primary.dark,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.overline.fontSize,
    textAnchor: 'middle'
  }
}))(RawChartAxes);

export default ChartAxes;
