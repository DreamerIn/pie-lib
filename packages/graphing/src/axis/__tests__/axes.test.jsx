import { shallow } from 'enzyme';
import React from 'react';
import { graphProps } from '../../__tests__/utils';

import { RawXAxis, RawYAxis } from '../axes';

describe('RawXAxis', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      graphProps: graphProps()
    };
    const props = { ...defaults, ...extras };
    return shallow(<RawXAxis {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });
});

describe('RawYAxis', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      graphProps: graphProps()
    };
    const props = { ...defaults, ...extras };
    return shallow(<RawYAxis {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });
});
