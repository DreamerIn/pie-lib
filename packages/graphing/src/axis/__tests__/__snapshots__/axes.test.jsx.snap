// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RawXAxis snapshot renders 1`] = `
<Fragment>
  <F
    hideZero={true}
    left={0}
    numTicks={1}
    scale={[MockFunction]}
    tickLabelProps={[Function]}
    top={0}
  />
  <WithStyles(Arrow)
    direction="left"
    scale={
      Object {
        "x": [MockFunction],
        "y": [MockFunction] {
          "calls": Array [
            Array [
              0,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": 0,
            },
          ],
        },
      }
    }
    x={0}
    y={0}
  />
  <WithStyles(Arrow)
    direction="right"
    scale={
      Object {
        "x": [MockFunction],
        "y": [MockFunction] {
          "calls": Array [
            Array [
              0,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": 0,
            },
          ],
        },
      }
    }
    x={1}
    y={0}
  />
</Fragment>
`;

exports[`RawYAxis snapshot renders 1`] = `
<Fragment>
  <F
    height={400}
    hideZero={true}
    left={0}
    numTicks={1}
    orientation="left"
    scale={[MockFunction]}
    tickLabelProps={[Function]}
    tickLength={10}
    tickTextAnchor="bottom"
    top={0}
  />
  <WithStyles(Arrow)
    direction="down"
    scale={
      Object {
        "x": [MockFunction] {
          "calls": Array [
            Array [
              0,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": 0,
            },
          ],
        },
        "y": [MockFunction],
      }
    }
    x={0}
    y={0}
  />
  <WithStyles(Arrow)
    direction="up"
    scale={
      Object {
        "x": [MockFunction] {
          "calls": Array [
            Array [
              0,
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": 0,
            },
          ],
        },
        "y": [MockFunction],
      }
    }
    x={0}
    y={1}
  />
</Fragment>
`;
